### ISSUE TITLE
mdata-put randomly hangs

### ISSUE THREAD

**Issue by blackwood821** [joyent/smartos-live#990]
We have an upgrade routine where after we reprovision a VM to a new image our `user-script` runs an upgrade process on the zone. Every once in a while our script fails because it runs an `mdata-put` command that times out. Here is an example outside of our script that recreates the issue:

```
[root@ds (us-west-agc:bnbntl) ~]# mdata-put set:state success
plat_recv timeout
receive timeout, resetting protocol...
plat_recv timeout
receive timeout, resetting protocol...
plat_recv timeout
receive timeout, resetting protocol...
plat_recv timeout
receive timeout, resetting protocol...
```

Restarting the `vminfod` service via `svcadm restart vminfod` on the CN global always seems to fix the issue.

There is more info available at https://gist.github.com/blackwood821/d343ee38d76a44f5ebb1429828787199.

**Comment by blackwood821** [comment]
Here is a recent error from the `metadata` service log in the global zone:
```
[2021-07-09T20:38:05.353Z]  INFO: metadata/752543 on cn01: handled GET setup:messages (brand=joyent, zonename=7d38ff30-de00-46af-9a8b-16c41c6220e7, response=SUCCESS)
    query: {
      "cmd": "GET",
      "arg": "setup:messages",
      "req_v": 2,
      "elapsed": 0.33317
    }
[2021-07-09T20:38:30.234Z] DEBUG: metadata/752543 on cn01: createServersOnExistingZones for 38 zones
[2021-07-09T20:38:30.234Z]  INFO: metadata/752543 on cn01: created zone metadata sockets on 0 / 38 zones took 0.000385944s (385.94us)
[2021-07-09T20:39:05.481Z] ERROR: metadata/752543 on cn01: could not set metadata (key "setup:messages") (brand=joyent, zonename=7d38ff30-de00-46af-9a8b-16c41c6220e7)
    MultiError: first of 1 error: first of 1 error: vminfod watchForEvent "VM.js writeAllMetadata (7d38ff30-de00-46af-9a8b-16c41c6220e7: customer_metadata,internal_metadata,tags)" timeout exceeded
        at /usr/vm/node_modules/vasync/lib/vasync.js:99:14
        at /usr/vm/node_modules/VM.js:16887:29
        at /usr/vm/node_modules/VM.js:17072:13
        at /usr/node/0.10/node_modules/async.js:240:13
        at /usr/node/0.10/node_modules/async.js:144:21
        at /usr/node/0.10/node_modules/async.js:237:17
        at /usr/node/0.10/node_modules/async.js:600:34
        at /usr/vm/node_modules/VM.js:17014:21
        at /usr/vm/node_modules/VM.js:3690:9
        at next (/usr/vm/node_modules/vasync/lib/vasync.js:843:6)
    Caused by: MultiError: first of 1 error: vminfod watchForEvent "VM.js writeAllMetadata (7d38ff30-de00-46af-9a8b-16c41c6220e7: customer_metadata,internal_metadata,tags)" timeout exceeded
        at /usr/vm/node_modules/vasync/lib/vasync.js:99:14
        at done (/usr/vm/node_modules/vminfod/client.js:697:9)
        at vminfodEventStreamWatchForEventTimeout (/usr/vm/node_modules/vminfod/client.js:635:13)
        at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)
    Caused by: Error: vminfod watchForEvent "VM.js writeAllMetadata (7d38ff30-de00-46af-9a8b-16c41c6220e7: customer_metadata,internal_metadata,tags)" timeout exceeded
        at vminfodEventStreamWatchForEventTimeout (/usr/vm/node_modules/vminfod/client.js:633:23)
        at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)
[2021-07-09T20:39:05.481Z] ERROR: metadata/752543 on cn01: first of 1 error: first of 1 error: vminfod watchForEvent "VM.js writeAllMetadata (7d38ff30-de00-46af-9a8b-16c41c6220e7: customer_metadata,internal_metadata,tags)" timeout exceeded (brand=joyent, zonename=7d38ff30-de00-46af-9a8b-16c41c6220e7)
[2021-07-09T20:39:05.481Z] ERROR: metadata/752543 on cn01: Socket for 7d38ff30-de00-46af-9a8b-16c41c6220e7 closed before we could write anything. (brand=joyent, zonename=7d38ff30-de00-46af-9a8b-16c41c6220e7)
[2021-07-09T20:39:05.481Z]  INFO: metadata/752543 on cn01: handled PUT setup:messages (brand=joyent, zonename=7d38ff30-de00-46af-9a8b-16c41c6220e7, response=FAILURE)
    MultiError: first of 1 error: first of 1 error: vminfod watchForEvent "VM.js writeAllMetadata (7d38ff30-de00-46af-9a8b-16c41c6220e7: customer_metadata,internal_metadata,tags)" timeout exceeded
        at /usr/vm/node_modules/vasync/lib/vasync.js:99:14
        at /usr/vm/node_modules/VM.js:16887:29
        at /usr/vm/node_modules/VM.js:17072:13
        at /usr/node/0.10/node_modules/async.js:240:13
        at /usr/node/0.10/node_modules/async.js:144:21
        at /usr/node/0.10/node_modules/async.js:237:17
        at /usr/node/0.10/node_modules/async.js:600:34
        at /usr/vm/node_modules/VM.js:17014:21
        at /usr/vm/node_modules/VM.js:3690:9
        at next (/usr/vm/node_modules/vasync/lib/vasync.js:843:6)
    Caused by: MultiError: first of 1 error: vminfod watchForEvent "VM.js writeAllMetadata (7d38ff30-de00-46af-9a8b-16c41c6220e7: customer_metadata,internal_metadata,tags)" timeout exceeded
        at /usr/vm/node_modules/vasync/lib/vasync.js:99:14
        at done (/usr/vm/node_modules/vminfod/client.js:697:9)
        at vminfodEventStreamWatchForEventTimeout (/usr/vm/node_modules/vminfod/client.js:635:13)
        at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)
    Caused by: Error: vminfod watchForEvent "VM.js writeAllMetadata (7d38ff30-de00-46af-9a8b-16c41c6220e7: customer_metadata,internal_metadata,tags)" timeout exceeded
        at vminfodEventStreamWatchForEventTimeout (/usr/vm/node_modules/vminfod/client.js:633:23)
        at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)
    --
    query: {
      "cmd": "PUT",
      "arg": "setup:messages",
      "req_v": 2,
      "elapsed": 60120.467837
    }

```

**Comment by blackwood821** [comment]
Bump

**Comment by blackwood821** [comment]
This is constantly an issue for us whenever we upgrade a stack of SmartOS zones to a new set of images. Can anyone look into this?

**Comment by danmcd** [comment]
I'm very sorry I didn't see this one sooner.

Next time you get a "hang" please, before trying other things, gather the following information so we know what sort of hang it is.  We had someone recently report an mdata-put hang that may turn out to be a hard-to-reach-but-bad-if-reached bug in sockfs. Generally this will be good advice for any process hang:

1.) See what pstack(1) has to say.  I'll show a suspended cat(1) as an example:

```
[root@nuc ~]# pstack `pgrep cat`
83950:  cat
 feee0f35 read     (0, 80632a0, 2000)
 08051cc5 cat      (806b2c0, 8047adc, 8047b6c, 8052819) + 7e
 080524ef main     (8047c1c, fef5b784, 8047c58, 805146b) + 458
 0805146b _start_crt (1, 8047c80, fefcec37, 0, 0, 0) + 9a
 0805133a _start   (1, 8047d8c, 0, 8047d90, 8047de3, 8047df7) + 1a
[root@nuc ~]# 
```
pstack(1) can be peformed by same-user-in-zone, or by root@global-zone. For multithreaded processes, pstack will produce a lot of output. Sometimes pstack will produce NO output, or no output for a single thread in a multithreaded process.  And that's a sign to move on to step 2:

2.) As root@global-zone, see if you can where the kernel is hanging:  `mdb -ke "0t<PID>::pid2proc |::walk thread |::findstack -v"`

```
[root@nuc ~]# mdb -ke "0t83950::pid2proc |::walk thread |::findstack -v"
stack pointer for thread fffffe59fa7f2b60 (cat/1): fffffe007ac808e0
[ fffffe007ac808e0 _resume_from_idle+0x12b() ]
  fffffe007ac80910 swtch+0x139()
  fffffe007ac80980 stop+0x2dd(5, 1a)
  fffffe007ac809b0 isjobstop+0xac(1a)
  fffffe007ac80a60 issig_forreal+0x1dc()
  fffffe007ac80a80 issig+0x15(0)
  fffffe007ac80b00 cv_wait_sig_swap_core+0x277(fffffffffbccccb0, fffffe58c8b23040, 0)
  fffffe007ac80b20 cv_wait_sig_swap+0x17(fffffffffbccccb0, fffffe58c8b23040)
  fffffe007ac80b90 straccess+0xe5(fffffe590f39b4f0, 0)
  fffffe007ac80c30 strread+0x1aa(fffffe58f07b5d80, fffffe007ac80da0, fffffe58ef25a4c0)
  fffffe007ac80cc0 spec_read+0x22b(fffffe58f07b5d80, fffffe007ac80da0, 0, fffffe58ef25a4c0, 0)
  fffffe007ac80d60 fop_read+0x115(fffffe58f07b5d80, fffffe007ac80da0, 0, fffffe58ef25a4c0, 0)
  fffffe007ac80e80 read+0x199(0, 80632a0, 2000)
  fffffe007ac80eb0 read32+0x1e(0, 80632a0, 2000)
  fffffe007ac80f10 _sys_sysenter_post_swapgs+0x159()
[root@nuc ~]# 
```

For multithreaded processes this will generate a block somewhere, usually at a kernel synchronization primitive like mutex_enter, cv_wait(), or related calls.

It's interesting that you were able to unstick things by restarting vminfod. I do wonder if, next time you encounter this, BEFORE restarting vminfod capturing some of those mdata-put stacks.  ALSO of interest, see if the OLD mdata-put processes are still lingering after you restart vminfod and get your system to a stable state.

**Comment by blackwood821** [comment]
@danmcd Thanks for the reply! We had another batch of customer upgrades and it happened several times. I put a watchdog script in place that captured a good amount of data based on your suggestions. I'll review and post it next week.

**Comment by chadschmelter** [comment]
@danmcd I captured this in production during one of our upgrades. Sharing the data and Claude analysis of what I think it shows.

## Setup

I wrote a watchdog that runs on the CN GZ. It polls once a second for any in-zone `mdata-*` process alive longer than 30 seconds, and when it finds one it snapshots: the in-zone process state (`pstack` + `pfiles` + kernel `findstack`), the metadata service state, vminfod state, `vminfod /status` JSON, `fuser` against every per-zone qlocker, and the last 500 lines of both service logs. It's read-only and doesn't run `gcore`, so it should be safe to leave running during an upgrade.

Six captures of the same zone over an 8-minute window. Full tarball attached.

## What the data shows

The failing zone is one of 79 on the CN. Other zones are completing `PUT setup:state` in **250-340 ms** in the same time window. Only this one zone is stuck.

**The metadata service is healthy.** It acquires the per-zone qlocker fcntl lock cleanly each cycle. The `metadata.json` rewrite via `writeAndRename` is succeeding (the file's `mtime` keeps advancing). No thread is blocked on the lock or anywhere else (main thread in `port_getn`, libuv workers parked in `cond_wait`). pfiles reports the lockfile fd as `advisory write lock set by process N`, where N is the metadata service itself, not a third-party stuck holder. So it's holding its own lock cleanly and waiting at the JS layer for `vs.watchForEvent` to fire.

**Vminfod is healthy.** Every snapshot shows `vminfod /status` reporting `state: running, status: working, queue.idle: true, queue.npending: 0, queue.nqueued: 0, paused: false`. No `refreshErrors`. All threads parked in `port_getn` / `lwp_park`. It's not overloaded and it's not stuck.

**But vminfod is not seeing this zone's metadata.json changes.** In the 500 lines of vminfod log around the failures, only **one** modify event was emitted for this zone, and it came from `self.reset()` (vminfod's periodic full refresh, which runs every 5 minutes), not from a filesystem event:

```
{"msg":"self.reset() took 0.616963420s (616.96ms): 1 vms changed",
 "changes":[
   {"path":["last_modified"],"oldValue":"02:41:08","newValue":"02:43:10"},
   {"path":["customer_metadata","setup:state"],"action":"added","newValue":"running"}]}
```

The PUT outcomes line up with whether `self.reset()` lands inside or outside the 60s `VMINFOD_TIMEOUT` window:

```
02:42:10 FAILURE elapsed=60188ms   (watchForEvent timed out)
02:43:36 SUCCESS elapsed=56651ms   ← self.reset() hit at 02:43:36, just inside the 60s
02:43:36 SUCCESS elapsed=11884ms   ← rode the same self.reset()
02:44:38 FAILURE elapsed=60172ms   (next mdata-put, self.reset() too far away)
```

## What I think this means

The per-VM fswatcher watch on `/zones/<uuid>/config/metadata.json` appears dead for this one zone. Vminfod is only catching the file's changes via the 5-minute `self.reset()`. An `mdata-put` only succeeds when a `self.reset()` happens to fire inside its 60-second `watchForEvent` timeout, which works out to roughly a 1-in-5 chance per attempt, so most attempts time out.

My best guess at the mechanism: every `writeAndRename` on `metadata.json` produces an event against the old inode that fswatcher.c flags as `is_final = TRUE` (per `port_associate(3C)`, `FILE_DELETE` and friends are delivered unconditionally on `PORT_SOURCE_FILE` watches when the watched file is unlinked or replaced). The C side destroys the watch handle and the JS side queues an async re-watch via stdin to the C process. There's a `LONG_WATCH_INTERVAL` in fswatcher.js that's supposed to keep retrying every 10 seconds for files that don't exist yet, so in principle the re-watch should land eventually. I haven't pinned down why it doesn't in this case. The most likely candidates are a race during reprovision when the zone root is briefly destroyed and recreated, or something in the C-side watch-command pipeline getting wedged, but I can't say for sure from the captures alone.

Once that happens, the watch appears dead for the life of the vminfod process. `svcadm restart vminfod` re-arms all watches and the symptom clears, which is the same workaround we've been using since 2021.

## Diagnostic recipe for the next reporter

When `mdata-put` hangs persistently from inside a zone, run on the GZ:

```sh
ZONE=<zone-uuid>

# 1. Confirm the metadata service is in the watchForEvent path (not stuck elsewhere)
pfiles $(pgrep -f /usr/vm/sbin/metadata) | \
    grep -B1 -A2 "${ZONE}.*config.lockfile"

# 2. Check vminfod is not actually busy
curl -s http://127.0.0.1:9090/status | json queue lastRefresh refreshErrors

# 3. Check whether vminfod's events for this zone are coming from self.reset()
#    rather than from fs events. If yes, the per-VM watch appears dead.
tail -1000 /var/svc/log/system-smartdc-vminfod:default.log | \
    grep "$ZONE" | grep -E 'self\.reset|emitting.*modify'
```

If you see modify events for the zone **only** in `self.reset()` log lines, the fswatcher watch appears dead and `svcadm restart vminfod` is the recovery.

Re your third question (whether old `mdata-put` processes linger after restarting vminfod): I didn't snapshot the post-restart state on this particular CN, but the in-zone `mdata-put` retry loops exited on their own once vminfod came back, with no observable orphans. I'll capture that explicitly the next time this fires.

[mdata-hangs-cn01-20260511-194526.tar.gz](https://github.com/user-attachments/files/27977461/mdata-hangs-cn01-20260511-194526.tar.gz)
